### DISCUSSION TITLE
humble-kirkstone Wiki Creation Questions

### DISCUSSION THREAD

**Discussion by Ryanf55** [bmwcarit/meta-ros#1093] [Q&A]
Hello! I'm creating a public discussion to get help resolving my questions when writing this tutorial: https://github.com/Ryanf55/poky/wiki/Manual-Setup-Process

I'm currently facing an issue with `bitbake-layers add-layer ../meta-ros/meta-ros2`, where it reports the error: `ERROR: Layer ros2-layer is not compatible with the core layer which only supports these series: kirkstone (layer is compatible with ${ROS_OE_RELEASE_SERIES})
`

It was recommended to switch meta-ros from `kirkstone` to `kirkstone-next`. I've done that, and meta-ros is now pointing at hash baf1fa6859fa6f60e970c13b97f7040a22246bb6. Unfortunately, I still get the same problem.

Rob also recommended following a gist. I think this is it. Should this line be `kirkstone` or `kirkstone-next`?  https://gist.github.com/robwoolley/e16f60e6e8483e7ea36f7616197bab11#file-build-ros-image-world-txt-L8

**Comment by robwoolley** [comment]
Hi Ryan,  

This should be the minimal configuration to get someone up and building with kirkstone-next.  The kirkstone-next branch should just be temporary measure for the short-term until we can verify it and merge it in to the kirkstone branch.

Thanks for taking up this project, it will be great for new users!

Regards,
Rob

```
git clone -b kirkstone git://git.yoctoproject.org/poky
```


> Cloning into 'poky'...
> remote: Enumerating objects: 633413, done.
> remote: Counting objects: 100% (393/393), done.
> remote: Compressing objects: 100% (311/311), done.
> remote: Total 633413 (delta 171), reused 137 (delta 76), pack-reused 633020
> Receiving objects: 100% (633413/633413), 201.26 MiB | 5.33 MiB/s, done.
> Resolving deltas: 100% (460962/460962), done.


```
git clone -b kirkstone https://git.openembedded.org/meta-openembedded
```

> Cloning into 'meta-openembedded'...
> remote: Enumerating objects: 230871, done.
> remote: Counting objects: 100% (400/400), done.
> remote: Compressing objects: 100% (207/207), done.
> remote: Total 230871 (delta 213), reused 317 (delta 183), pack-reused 230471
> Receiving objects: 100% (230871/230871), 57.60 MiB | 19.95 MiB/s, done.
> Resolving deltas: 100% (140694/140694), done. 

```
git clone -b kirkstone-next https://github.com/ros/meta-ros.git
```

> Cloning into 'meta-ros'...
> remote: Enumerating objects: 141711, done.
> remote: Counting objects: 100% (44247/44247), done.
> remote: Compressing objects: 100% (6632/6632), done.
> remote: Total 141711 (delta 37849), reused 42432 (delta 37023), pack-reused 97464
> Receiving objects: 100% (141711/141711), 38.76 MiB | 13.39 MiB/s, done.
> Resolving deltas: 100% (102465/102465), done.
> Updating files: 100% (15513/15513), done.



```
source oe-init-build-env
```

> You had no conf/local.conf file. This configuration file has therefore been
> created for you from /ala-lpggp31/rwoolley/ryan/poky/meta-poky/conf/local.conf.sample
> You may wish to edit it to, for example, select a different MACHINE (target
> hardware). See conf/local.conf for more information as common configuration
> options are commented.
> 
> You had no conf/bblayers.conf file. This configuration file has therefore been
> created for you from /ala-lpggp31/rwoolley/ryan/poky/meta-poky/conf/bblayers.conf.sample
> To add additional metadata layers into your configuration please add entries
> to conf/bblayers.conf.
> 
> The Yocto Project has extensive documentation about OE including a reference
> manual which can be found at:
>     https://docs.yoctoproject.org/
> 
> For more information about OpenEmbedded see the website:
>     https://www.openembedded.org/
> 
> 
> ### Shell environment set up for builds. ###
> 
> You can now run 'bitbake <target>'
> 
> Common targets are:
>     core-image-minimal
>     core-image-full-cmdline
>     core-image-sato
>     core-image-weston
>     meta-toolchain
>     meta-ide-support
> 
> You can also run generated qemu images with a command like 'runqemu qemux86'
> 
> Other commonly useful commands are:
>  - 'devtool' and 'recipetool' handle common recipe tasks
>  - 'bitbake-layers' handles common layer tasks
>  - 'oe-pkgdata-util' handles common target package tasks



```
bitbake-layers add-layer ../meta-openembedded/meta-oe/
```
> NOTE: Starting bitbake server...

```
bitbake-layers add-layer ../meta-openembedded/meta-python/
```
> NOTE: Starting bitbake server...

```
bitbake-layers add-layer ../meta-ros/meta-ros-common/
```
> NOTE: Starting bitbake server...

```
bitbake-layers add-layer ../meta-ros/meta-ros2
```
> NOTE: Starting bitbake server...

```
bitbake-layers add-layer ../meta-ros/meta-ros2-humble
```
> NOTE: Starting bitbake server...

```
bitbake ros-image-core
```

**Reply by Ryanf55** [reply]
Thanks. Just needed an additional `cd poky` and then add an extra `../` to all the bitbake-layers commands above and it builds. 

Instead of manually cloning all the repos with `git clone`, it's standard in ROS to use [vcstool](https://github.com/dirk-thomas/vcstool). I think it would be a good and familiar way to conver the above `git clone` commands to use a vcs file.

The next step would be to run the image and inspect the files to find the ROS installation in `/opt/ros/humble` or wherever it shows up. 

I tried: `runqemu qemux86-64 ros-image-core`, but it failed. In the config, `MACHINE` is already set to `qemux86-64`.
```bash
ryan@B650-970:~/Dev/yocto_ws/poky/build$ bitbake ros-image-core
Loading cache: 100% |                                                                                                                                                                                                                                              | ETA:  --:--:--
Loaded 0 entries from dependency cache.
WARNING: /home/ryan/Dev/yocto_ws/poky/meta/recipes-kernel/lttng/babeltrace_1.5.11.bb: distutils-common-base.bbclass is deprecated, please use setuptools3-base.bbclass instead                                                                                      | ETA:  0:00:13
Parsing recipes: 100% |#############################################################################################################################################################################################################################################| Time: 0:00:08
Parsing of 3762 .bb files complete (0 cached, 3762 parsed). 5481 targets, 224 skipped, 1 masked, 0 errors.
NOTE: Resolving any missing task queue dependencies

Build Configuration:
BB_VERSION           = "2.0.0"
BUILD_SYS            = "x86_64-linux"
NATIVELSBSTRING      = "ubuntu-22.04"
TARGET_SYS           = "x86_64-poky-linux"
MACHINE              = "qemux86-64"
DISTRO               = "poky"
DISTRO_VERSION       = "4.0.15"
TUNE_FEATURES        = "m64 core2"
TARGET_FPU           = ""
DISTRO_NAME          = "Poky (Yocto Project Reference Distro)"
ROS_DISTRO           = "humble"
ROS_VERSION          = "2"
ROS_PYTHON_VERSION   = "3"
meta                 
meta-poky            
meta-yocto-bsp       = "my-kirkstone:f754b5d45b504e7486364ea9cfcb5f96086084bc"
meta-oe              
meta-python          = "kirkstone:730e44900a0a86265bad93a16b5a5ff344a07266"
meta-ros-common      
meta-ros2            
meta-ros2-humble     = "kirkstone-next:685215f9837d47089d2eaaa296c1dac574f44ddf"

NOTE: Fetching uninative binary shim http://downloads.yoctoproject.org/releases/uninative/4.3/x86_64-nativesdk-libc-4.3.tar.xz;sha256sum=1c35f09a75c4096749bbe1e009df4e3968cde151424062cf4aa3ed89db22b030 (will check PREMIRRORS first)
Initialising tasks: 100% |##########################################################################################################################################################################################################################################| Time: 0:00:01
Sstate summary: Wanted 2748 Local 0 Mirrors 0 Missed 2748 Current 0 (0% match, 0% complete)
NOTE: Executing Tasks
WARNING: libsdl2-native-2.0.20-r0 do_fetch: Failed to fetch URL http://www.libsdl.org/release/SDL2-2.0.20.tar.gz, attempting MIRRORS if available
WARNING: gtest-vendor-native-1.10.9004-4-r0 do_populate_lic: QA Issue: gtest-vendor-native: No generic license file exists for: BSD in any provider [license-exists]
WARNING: gmock-vendor-native-1.10.9004-4-r0 do_populate_lic: QA Issue: gmock-vendor-native: No generic license file exists for: BSD in any provider [license-exists]
WARNING: unique-identifier-msgs-2.2.1-3-r0 do_populate_lic: QA Issue: unique-identifier-msgs: No generic license file exists for: BSD in any provider [license-exists]
WARNING: pybind11-vendor-2.4.2-1-r0 do_populate_lic: QA Issue: pybind11-vendor: No generic license file exists for: BSD in any provider [license-exists]
WARNING: console-bridge-vendor-1.4.1-1-r0 do_populate_lic: QA Issue: console-bridge-vendor: No generic license file exists for: BSD in any provider [license-exists]
WARNING: class-loader-2.2.0-3-r0 do_populate_lic: QA Issue: class-loader: No generic license file exists for: BSD in any provider [license-exists]
WARNING: pluginlib-5.1.0-3-r0 do_populate_lic: QA Issue: pluginlib: No generic license file exists for: BSD in any provider [license-exists]
WARNING: ros-image-core-1.0-r0 do_rootfs: QA Issue: The license listed BSD was not in the licenses collected for recipe class-loader [license-file-missing]
WARNING: ros-image-core-1.0-r0 do_rootfs: QA Issue: The license listed BSD was not in the licenses collected for recipe console-bridge-vendor [license-file-missing]
WARNING: ros-image-core-1.0-r0 do_rootfs: QA Issue: The license listed BSD was not in the licenses collected for recipe pluginlib [license-file-missing]
WARNING: ros-image-core-1.0-r0 do_rootfs: QA Issue: The license listed BSD was not in the licenses collected for recipe unique-identifier-msgs [license-file-missing]
NOTE: Tasks Summary: Attempted 6903 tasks of which 0 didn't need to be rerun and all succeeded.

Summary: There were 13 WARNING messages.
ryan@B650-970:~/Dev/yocto_ws/poky/build$ runqemu qemux86-64 ros-image-core
runqemu - INFO - Running MACHINE=qemux86-64 bitbake -e ...
runqemu - ERROR - Failed to find /home/ryan/Dev/yocto_ws/poky/build/tmp/deploy/images/qemux86-64/ros-image-core-qemux86-64.qemuboot.conf (wrong image name or BSP does not support running under qemu?).
runqemu - INFO - Cleaning up
runqemu - INFO - Host uptime: 370845.95

```

Any steps on what to do to run the image with qemu to get a shell? There are a bunch of other seemingly related files in `poky/build/tmp/deploy/images/qemux86-64`

**Reply by robwoolley** [reply]
I have used the vctool.  My only suggestion would be to double check that it is being maintained going forward.  As far as I know it hasn't been updated in 3 years and may no longer be actively maintained.

It looks like the files have the ROS distro appended to the name.  (This appears to happen in meta-ros-common/classes/ros_image.bbclass: 
```
# Allow images of different ROS_DISTRO-s to be distinguished.
ROS_IMAGE_BASENAME_APPEND ??= "-${ROS_DISTRO}"
IMAGE_BASENAME:append = "${ROS_IMAGE_BASENAME_APPEND}"
```
So, there is an extra "humble" in the name compared to what runqemu expects:
```
$ ls poky/build/tmp/deploy/images/qemux86-64
bzImage
bzImage--5.15.124+git0+f484a7f175_1c09be01f4-r0-qemux86-64-20240123112958.bin
bzImage-qemux86-64.bin
modules--5.15.124+git0+f484a7f175_1c09be01f4-r0-qemux86-64-20240123112958.tgz
modules-qemux86-64.tgz
ros-image-core-humble-qemux86-64-20240123112958.qemuboot.conf
ros-image-core-humble-qemux86-64-20240123112958.rootfs.ext4
ros-image-core-humble-qemux86-64-20240123112958.rootfs.manifest
ros-image-core-humble-qemux86-64-20240123112958.rootfs.tar.bz2
ros-image-core-humble-qemux86-64-20240123112958.testdata.json
ros-image-core-humble-qemux86-64.ext4
ros-image-core-humble-qemux86-64.manifest
ros-image-core-humble-qemux86-64.qemuboot.conf
ros-image-core-humble-qemux86-64.tar.bz2
ros-image-core-humble-qemux86-64.testdata.json
```

I think you should be able to just change the command to:
```
runqemu qemux86-64 ros-image-core-humble
```

**Reply by Ryanf55** [reply]
Excellent, that works. I've updated the wiki with the instructions.

I'm not too worried, vcstool is working fine as-is, used in the ROS build farm, and OSRF has said they will fix bugs. If we hit any bugs using it for the tutorial for the working group, I'm sure we could find someone to help. There's also [vcstool2](https://pypi.org/project/vcstool2/) which has some improvements. I've been using the original vcstool at work and for ArduPilot with zero issues. The template repository could easily add support for other tools like Google Repo.

Here's the equivalent repos file:

**ros-image-core.repos**
```yaml
repositories:
  poky:
    type: git
    url: git://git.yoctoproject.org/poky
    version: kirkstone
  meta-openembedded:
    type: git
    url: https://git.openembedded.org/meta-openembedded
    version: kirkstone
  meta-ros:
    type: git
    url: https://github.com/ros/meta-ros.git
    # TODO switch to kirkstone once kirkstone-next is merged
    version: kirkstone-next
```

Which simplifies instructions to:
```bash
# Import sources with vcstool. Other tools like Google Repo could also be supported following the same format
vcs import --input ros-image-core.repos
# Initialize the build environment
cd poky
source oe-init-build-env
# Add all the layers
bitbake-layers add-layer ../../meta-openembedded/meta-oe/
bitbake-layers add-layer ../../meta-openembedded/meta-python/
bitbake-layers add-layer ../../meta-ros/meta-ros-common/
bitbake-layers add-layer ../../meta-ros/meta-ros2
bitbake-layers add-layer ../../meta-ros/meta-ros2-humble
# Run the build
bitbake ros-image-core
# Run the image
runqemu qemux86-64 ros-image-core-humble
```

**Next Question:**
While I could get passed the sudo prompt for the `tap` interface, the next problem was that it's asking for username/password. My system credentials didn't work. Is there a default user/password for openembedded? I tried seaching the poky docs for "password" but didn't find anything related. 
![image](https://github.com/ros/meta-ros/assets/25047695/a00db590-7f91-445b-912d-8ce7580199fc)

**Comment by robwoolley** [comment]
Hi Ryan,

By default the poky builds have "debug-tweaks" enabled in conf/local.conf.  In this debug mode, the root password is blank.  So you can just log in with root and you will not get prompted for a password:

> Poky (Yocto Project Reference Distro) 4.0.15 qemux86-64 /dev/ttyS0
> 
> qemux86-64 login: root
> root@qemux86-64:\~#


When I tested it just now, I was using it on a build server where I was connected via SSH and didn't have root privileges.  Adding slirp and nographic to runqemu allowed me to run it in the console without root privileges for the tap devices.  I did find however, that I had to manually specify the cpu type as it had defaulted to compile for the IvyBridge processor and I was getting complaints unsupported CPUID features (from QEMU):

> runqemu - ERROR - Failed to run qemu: qemu-system-x86_64: warning: TCG doesn't support requested feature: CPUID.01H:ECX.x2apic [bit 21]                                                                     
> qemu-system-x86_64: warning: TCG doesn't support requested feature: CPUID.01H:ECX.tsc-deadline [bit 24]
> qemu-system-x86_64: warning: TCG doesn't support requested feature: CPUID.01H:ECX.avx [bit 28]        
> qemu-system-x86_64: warning: TCG doesn't support requested feature: CPUID.01H:ECX.f16c [bit 29]       
> qemu-system-x86_64: warning: TCG doesn't support requested feature: CPUID.01H:ECX.x2apic [bit 21]     
> qemu-system-x86_64: warning: TCG doesn't support requested feature: CPUID.01H:ECX.tsc-deadline [bit 24]
> qemu-system-x86_64: warning: TCG doesn't support requested feature: CPUID.01H:ECX.avx [bit 28]        
> qemu-system-x86_64: warning: TCG doesn't support requested feature: CPUID.01H:ECX.f16c [bit 29]      
>  qemu-system-x86_64: warning: TCG doesn't support requested feature: CPUID.01H:ECX.x2apic [bit 21]     
> qemu-system-x86_64: warning: TCG doesn't support requested feature: CPUID.01H:ECX.tsc-deadline [bit 24]
> qemu-system-x86_64: warning: TCG doesn't support requested feature: CPUID.01H:ECX.avx [bit 28]        
> qemu-system-x86_64: warning: TCG doesn't support requested feature: CPUID.01H:ECX.f16c [bit 29]       
> qemu-system-x86_64: warning: TCG doesn't support requested feature: CPUID.01H:ECX.x2apic [bit 21]     
> qemu-system-x86_64: warning: TCG doesn't support requested feature: CPUID.01H:ECX.tsc-deadline [bit 24]
> qemu-system-x86_64: warning: TCG doesn't support requested feature: CPUID.01H:ECX.avx [bit 28]        
> qemu-system-x86_64: warning: TCG doesn't support requested feature: CPUID.01H:ECX.f16c [bit 29]       
> Could not initialize SDL(x11 not available) - exiting  

                                               

 or exceptions at runtime for unimplemented instructions if I chose an earlier CPU variant:

> root@qemux86-64:\~# ros2
> [  107.455050] traps: python3[328] trap invalid opcode ip:7f6d9f5dbbea sp:7ffcbae7b8f0 error:0 in lib]
> Illegal instruction

**Issue 1:** The QEMU -cpu option seems like a possible bug in the Yocto QEMU BSP because the machine is configured to build for the IvyBridge instruction set, but it hasn't seemingly set the variables used by runqemu to match.  This perhaps isn't seen when using kvm.

I temporarily worked around this with:
```
runqemu qemux86-64 qemuparams="-cpu IvyBridge" ros-image-core-humble slirp nographic
```

I was able to log in and source /usr/bin/ros_setup.sh.  However it complained about unknown parameters to "head".  This is because poky has busybox by default for constrained devices, however ROS makes some assumptions about bash and coreutils being installed.

**Issue 2:** I tried sourcing the ros_setup.sh to source the environment, but I found that it required "head" from coreutils instead of the minimalistic version found in busybox.

The proper thing would be for us to fix the ros-image-core.bb image recipe, (I added a comment to #1068) however as a quick fix we can just add these lines to conf/local.conf and rebuild ros-image-core:

```
# the deploy code requires bash and
# normal linux utilities not busybox ones
CORE_IMAGE_EXTRA_INSTALL += "\
    bash coreutils util-linux tar gzip bzip2 kmod \
    python3-modules python3-misc \
    e2fsprogs e2fsprogs-mke2fs parted \
    "
```

With the new image we can then source ros_setup.sh and run ros2. The ros2 command is a little slow to start and may take 10 to 20 seconds to produce output. (This is something to investigate as well)

> Poky (Yocto Project Reference Distro) 4.0.15 qemux86-64 ttyS0
> 
> qemux86-64 login: root
> root@qemux86-64:\~# echo $SHELL
> /bin/sh
> root@qemux86-64:\~# ls -al /bin/sh
> lrwxrwxrwx 1 root root 14 Mar  9  2018 /bin/sh -> /bin/bash.bash
> root@qemux86-64:\~# source /usr/bin/ros_setup.sh
> root@qemux86-64:\~# ros2
> usage: ros2 [-h] [--use-python-default-buffering]
>             Call `ros2 <command> -h` for more detailed usage. ...
> 
> ros2 is an extensible command-line tool for ROS 2.
> 
> options:
>   -h, --help            show this help message and exit
>   --use-python-default-buffering
>                         Do not force line buffering in stdout and instead use the python default
>                         buffering, which might be affected by PYTHONUNBUFFERED/-u and depends on
>                         whatever stdout is interactive or not
> 
> Commands:
>   action     Various action related sub-commands
>   component  Various component related sub-commands
>   daemon     Various daemon related sub-commands
>   doctor     Check ROS setup and other potential issues
>   interface  Show information about ROS interfaces
>   launch     Run a launch file
>   lifecycle  Various lifecycle related sub-commands
>   multicast  Various multicast related sub-commands
>   node       Various node related sub-commands
>   param      Various param related sub-commands
>   pkg        Various package related sub-commands
>   run        Run a package specific executable
>   security   Various security related sub-commands
>   service    Various service related sub-commands
>   topic      Various topic related sub-commands
>   wtf        Use `wtf` as alias to `doctor`
> 
>   Call `ros2 <command> -h` for more detailed usage.

Hope this helps!

**Reply by Ryanf55** [reply]
Great! I followed all that. 

I just filed https://github.com/ros/meta-ros/issues/1094 for the slow ROS 2 CLI. Not a blocker. I'll get this all added to the wiki.

**Reply by robwoolley** [reply]
Excellent, thank you.

**Reply by Ryanf55** [reply]
And, adding bash (and others) to ros-image-core PR: https://github.com/ros/meta-ros/pull/1095
